File: packages/uns-core/src/uns-mqtt/uns-mqtt-proxy.ts
import { readFileSync } from "fs";
import { IClientPublishOptions } from "mqtt";
import * as path from "path";
import { Worker } from "worker_threads";
import { fileURLToPath } from "url";
import { basePath } from "../base-path.js";
import logger from "../logger.js";
import { IMqttMessage, IUnsPacket, IUnsParameters, UnsEvents, ValueType } from "../uns/uns-interfaces.js";
import { MeasurementUnit } from "../uns/uns-measurements.js";
import { UnsPacket } from "../uns/uns-packet.js";
import { IMqttParameters, IMqttWorkerData } from "./mqtt-interfaces.js";
import { MqttTopicBuilder } from "./mqtt-topic-builder.js";
import UnsProxy from "../uns/uns-proxy.js";
import { UnsAttributeType } from "../graphql/schema.js";

const packageJsonPath = path.join(basePath, "package.json");
const packageJson = JSON.parse(readFileSync(packageJsonPath, "utf8"));

const moduleDirectory = path.dirname(fileURLToPath(import.meta.url));
const packageRoot = path.resolve(moduleDirectory, "..", "..");
const workerScriptPath = path.join(packageRoot, "dist/uns-mqtt/mqtt-worker-init.js");

export enum MessageMode {
  Raw = 'raw',     // Send only the original message
  Delta = 'delta', // Send only the delta message
  Both = 'both'    // Send both the original and delta messages
}

export default class UnsMqttProxy extends UnsProxy {
  private lastValues: Map<string, { value: ValueType; uom: string; timestamp: Date }> = new Map();
  private worker: Worker;
  private pendingEnqueues: Map<string, { resolve: () => void; reject: (reason?: any) => void }> = new Map();
  private unsParameters: IUnsParameters;
  protected processStatusTopic: string;
  public instanceName: string;
  private currentSequenceId: Map<string, number> = new Map();
  private topicBuilder: MqttTopicBuilder;

  constructor(
    mqttHost: string,
    processName: string,
    instanceName: string,
    unsParameters?: IUnsParameters,
    publisherActive: boolean = false,
    subscriberActive: boolean = false
  ) {
    super();
    this.instanceName = instanceName;
    // Create the topic builder using packageJson values and the processName.
    this.topicBuilder = new MqttTopicBuilder(`uns-infra/${MqttTopicBuilder.sanitizeTopicPart(packageJson.name)}/${MqttTopicBuilder.sanitizeTopicPart(packageJson.version)}/${MqttTopicBuilder.sanitizeTopicPart(processName)}/`);

    // Generate the processStatusTopic using the builder.
    this.processStatusTopic = this.topicBuilder.getProcessStatusTopic();
    // Derive the instanceStatusTopic by appending the instance name.
    this.instanceStatusTopic = this.processStatusTopic + instanceName + "/";

    // Concatenate processName with instanceName for the worker identification.
    this.instanceNameWithSuffix = `${processName}-${instanceName}`;
    
    const mqttParameters: IMqttParameters = {
      mqttSubToTopics: unsParameters?.mqttSubToTopics ?? [],
      username: unsParameters?.username ?? "",
      password: unsParameters?.password ?? "",
      mqttSSL: unsParameters?.mqttSSL ?? false,
      statusTopic: this.instanceStatusTopic,
    };
    this.unsParameters = unsParameters ?? {};
    this.startQueueWorker(mqttHost, this.instanceNameWithSuffix, mqttParameters, publisherActive, subscriberActive);
  }

  /**
   * Starts a worker thread to process the throttled publish queue.
   */
  private startQueueWorker(
    mqttHost: string,
    instanceNameWithSuffix: string,
    mqttParameters: IMqttParameters,
    publisherActive: boolean,
    subscriberActive: boolean
  ): void {
    const workerData: IMqttWorkerData = {
      publishThrottlingDelay: this.unsParameters.publishThrottlingDelay ?? 1,
      subscribeThrottlingDelay: this.unsParameters.subscribeThrottlingDelay ?? 1,
      persistToDisk: false,
      mqttHost: mqttHost,
      instanceNameWithSuffix: instanceNameWithSuffix,
      mqttParameters: mqttParameters,
      publisherActive,
      subscriberActive
    };

    this.worker = new Worker(workerScriptPath, { workerData });

    this.worker.on("message", (msg) => {
      if (msg && msg.command === "enqueueResult" && msg.id) {
        const pending = this.pendingEnqueues.get(msg.id);
        if (pending) {
          if (msg.status === "success" && msg.topic && msg.message) {
            pending.resolve();
          } else {
            pending.reject(new Error(msg.error));
          }
          this.pendingEnqueues.delete(msg.id);
        }
      } else if (msg && msg.command === "input") {
        this.event.emit("input", { topic: msg.topic, message: msg.message.toString(), packet: msg.packet });
      } else if (msg && (msg.command === "handover_subscriber" || msg.command === "handover_publisher")) {
        this.event.emit("mqttWorker", { command: msg.command, batchSize: msg.batchSize, referenceHash: msg.referenceHash, instanceName: this.instanceName });
      } else if (msg && msg.command === "mqttProxyStatus") {
        this.event.emit("mqttProxyStatus", { event: msg.event, value: msg.value, uom: msg.uom, statusTopic: msg.statusTopic });
      }
    });

    this.worker.on("error", (err) => {
      logger.error("Error in worker:", err);
    });

    this.worker.on("exit", (code) => {
      if (code !== 0) {
        logger.error(`Worker exited with code ${code}`);
      }
    });
  }

  /**
   * Enqueues a message to the worker queue.
   *
   * @param topic - The topic to which the message belongs.
   * @param message - The message to be enqueued.
   * @param options - Optional publish options.
   * @returns A promise that resolves when the message is successfully enqueued.
   */
  private async enqueueMessageToWorkerQueue(topic: string, message: string, options?: IClientPublishOptions): Promise<void> {
    return new Promise((resolve, reject) => {
      // const id: string = String(this.currentSequenceId.get(topic) ?? 0);
      const id = `${Date.now()}-${Math.random()}`;
      this.pendingEnqueues.set(id, { resolve, reject });
      this.worker.postMessage({ command: "enqueue", id, topic, message, options });
    });
  }

  /**
   * Sets the publisher active state.
   *
   * @param batchSize - Optional batch size.
   * @param referenceHash - Optional reference hash.
   */
  public setPublisherActive(batchSize?: number, referenceHash?: string): void {
    this.worker.postMessage({ command: "setPublisherActive", batchSize, referenceHash });
  }

  /**
   * Sets the publisher to passive mode.
   * @returns A promise that resolves when the publisher is set to passive.
   */
  public setPublisherPassive(): Promise<UnsEvents["mqttWorker"]> {
    this.worker.postMessage({ command: "setPublisherPassive"});
    return new Promise((resolve) => {
      this.event.on("mqttWorker", (msg: UnsEvents["mqttWorker"]) => {
        if (msg.command === "handover_publisher") {
          logger.info(`${this.instanceNameWithSuffix} - Publisher set to passive.`);
          resolve(msg);
        }
      });
    });
  }

  /**
   * Sets the subscriber active state.
   *
   * @param batchSize - Optional batch size.
   * @param referenceHash - Optional reference hash.
   */
  public setSubscriberActive(batchSize?: number, referenceHash?: string): void {
    this.worker.postMessage({ command: "setSubscriberActive", batchSize, referenceHash });
  }

  /**
   * Sets the subscriber to passive mode.
   * @returns A promise that resolves when the subscriber is set to passive.
   */
  public setSubscriberPassive(): Promise<UnsEvents["mqttWorker"]> {
    this.worker.postMessage({ command: "setSubscriberPassive"});
    return new Promise((resolve) => {
      this.event.on("mqttWorker", (msg: UnsEvents["mqttWorker"]) => {
        if (msg.command === "handover_subscriber") {
          logger.info(`${this.instanceNameWithSuffix} - Publisher set to passive.`);
          resolve(msg);
        }
      });
    });
  }


  /**
   * Sets the subscriber to passive mode and allows the publisher to run
   * until the queue is empty (all messages are processed).
   */
  public async setSubscriberPassiveAndDrainQueue(): Promise<UnsEvents["mqttWorker"]> {
    return new Promise(async (resolve) => {
      const mqttWorkerData = await this.setSubscriberPassive();
      while (this.pendingEnqueues.size > 0) {
        await new Promise((resolve) => setTimeout(resolve, 100)); // Poll every 100ms
      }
      logger.info(`${this.instanceNameWithSuffix} - Subscriber set to passive and queue drained.`);
      resolve(mqttWorkerData);
    });
  }

  /**
   * Processes and publishes MQTT messages based on the selected message mode.
   *
   * @param mqttMessage - The MQTT message object.
   * @param mode - The message mode (Raw, Delta, or Both).
   */
  public publishMqttMessage(mqttMessage: IMqttMessage | null, mode: MessageMode = MessageMode.Raw) {
    if (mqttMessage) {
      if (mqttMessage.packet) {
        const time = UnsPacket.formatToISO8601(new Date());
        switch (mode) {
          case MessageMode.Raw: {
            this.processAndEnqueueMessage(mqttMessage, time, false);
            break;
          }
          case MessageMode.Delta: {
            const deltaMessage = { ...mqttMessage };
            deltaMessage.attribute = `${mqttMessage.attribute}-delta`;
            deltaMessage.description = `${mqttMessage.description} (delta)`;
            this.processAndEnqueueMessage(deltaMessage, time, true);
            break;
          }
          case MessageMode.Both: {
            this.processAndEnqueueMessage(mqttMessage, time, false);
            const deltaMessageBoth = { ...mqttMessage };
            deltaMessageBoth.attribute = `${mqttMessage.attribute}-delta`;
            deltaMessageBoth.description = `${mqttMessage.description} (delta)`;
            this.processAndEnqueueMessage(deltaMessageBoth, time, true);
            break;
          }
        }  
      } else {
        logger.error(`${this.instanceNameWithSuffix} - Error publishing mqtt message: mqttMessage.packet must be defined.`);
      }
    } else {
      logger.error(`${this.instanceNameWithSuffix} - Error publishing mqtt message: mqttMessage must be defined.`);
    }
  }

  /**
   * Publishes a message to a specified topic.
   *
   * @param topic - The MQTT topic.
   * @param message - The message to publish.
   * @returns A promise that resolves when enqueued.
   */
  public publishMessage(topic: string, message: string): Promise<void> {
    return this.enqueueMessageToWorkerQueue(topic, message);
  }

  /**
   * Parses an MQTT packet from a JSON string.
   *
   * @param mqttPacket - The MQTT packet string.
   * @returns A parsed IUnsPacket object or null.
   */
  public parseMqttPacket(mqttPacket: string): IUnsPacket | null {
    return UnsPacket.parseMqttPacket(mqttPacket, this.instanceNameWithSuffix);
  }

  /**
   * Subscribes asynchronously to one or more topics.
   *
   * @param topics - A topic or list of topics.
   */
  public subscribeAsync(topics: string | string[]): void {
    this.worker.postMessage({ command: "subscribeAsync", topics });
  }

  /**
   * Unsubscribes asynchronously from the given topics.
   *
   * @param topics - A list of topics.
   */
  public unsubscribeAsync(topics: string[]): void {
    this.worker.postMessage({ command: "unsubscribeAsync", topics });
  }

  /**
   * Processes and enqueues a message to the worker queue, including handling
   * sequencing, value differences, and tracking of unique topics.
   *
   * @param msg - The MQTT message to process.
   * @param time - The timestamp.
   * @param valueIsCumulative - Whether the value is cumulative.
   */
  private async processAndEnqueueMessage(msg: IMqttMessage, time: string, valueIsCumulative: boolean = false): Promise<void> {
    try {
      const attributeType =
        msg.packet.message.data ? UnsAttributeType.Data :
        msg.packet.message.event ? UnsAttributeType.Event :
        msg.packet.message.table ? UnsAttributeType.Table : null;
      
      let dataGroup = "";
      if (attributeType == UnsAttributeType.Data)
        dataGroup = msg.packet.message.data.dataGroup ?? "";
      if (attributeType == UnsAttributeType.Table)
        dataGroup = msg.packet.message.table.dataGroup ?? "";
      if (attributeType == UnsAttributeType.Event)
        dataGroup = msg.packet.message.event.dataGroup ?? "";

      this.registerUniqueTopic({
        timestamp: time,
        topic: msg.topic,
        attribute: msg.attribute,
        attributeType: attributeType,
        description: msg.description,
        tags: msg.tags,
        attributeNeedsPersistence: msg.attributeNeedsPersistence,
        dataGroup
      });

      const fullTopic = `${msg.topic}${msg.attribute}`;
      const sequenceId = this.currentSequenceId.get(msg.topic) ?? 0;
      this.currentSequenceId.set(msg.topic, sequenceId + 1);
      msg.packet.sequenceId = sequenceId;

      if (msg.packet.message.data) {
        const newValue = msg.packet.message.data.value;
        const newUom: MeasurementUnit = msg.packet.message.data.uom;
        const lastValueEntry = this.lastValues.get(fullTopic);
        const currentTime = new Date(msg.packet.message.data.time);

        if (lastValueEntry) {
          const intervalBetweenMessages = currentTime.getTime() - lastValueEntry.timestamp.getTime();
          const lastValue = lastValueEntry.value;
          this.lastValues.set(fullTopic, { value: newValue, uom: newUom, timestamp: currentTime });
          // Compute the delta and manage cumulative resets
          if (valueIsCumulative == true && typeof newValue === "number" && typeof lastValue === "number") {
            // Skip if newValue is 0 (likely a glitch)
            if (newValue === 0) {
              return; // Don't process or enqueue
            }
            const delta = newValue - lastValue;
            msg.packet.message.data.value = delta < 0 ? newValue : delta;
          }
          msg.packet.interval = intervalBetweenMessages;
          await this.enqueueMessageToWorkerQueue(fullTopic, JSON.stringify(msg.packet));
        } else {
          this.lastValues.set(fullTopic, { value: newValue, uom: newUom, timestamp: currentTime });
          logger.debug(`${this.instanceNameWithSuffix} - Need one more packet to calculate interval on topic ${fullTopic}`);
          if (valueIsCumulative === false) {
            await this.enqueueMessageToWorkerQueue(fullTopic, JSON.stringify(msg.packet));
          } else {
            logger.debug(`${this.instanceNameWithSuffix} - Need one more packet to calculate difference on value in data for topic ${fullTopic}`);
          }
        }
      } else if (msg.packet.message.command) {
        await this.enqueueMessageToWorkerQueue(fullTopic, JSON.stringify(msg.packet));
      } else if (msg.packet.message.event) {
        await this.enqueueMessageToWorkerQueue(fullTopic, JSON.stringify(msg.packet));
      } else if (msg.packet.message.table) {
        await this.enqueueMessageToWorkerQueue(fullTopic, JSON.stringify(msg.packet));
      }
    } catch (error: any) {
      logger.error(`${this.instanceNameWithSuffix} - Error publishing message to topic ${msg.topic}${msg.attribute}: ${error.message}`);
    }
  }

  /**
   * Stops the UnsProxy instance and cleans up resources.
   */
  public async stop(): Promise<void> {
    super.stop();
    // Terminate the worker thread if it exists.
    if (this.worker) {
      try {
        const exitCode = await this.worker.terminate();
        logger.info(`${this.instanceNameWithSuffix} - Worker terminated with exit code ${exitCode}`);
      } catch (error: any) {
        logger.error(`${this.instanceNameWithSuffix} - Error terminating worker: ${error.message}`);
      }
    }
    
    // Optionally, handle any pending enqueues.
    for (const [id, pending] of this.pendingEnqueues) {
      pending.reject(new Error("UnsProxy has been stopped"));
      this.pendingEnqueues.delete(id);
    }
  }

}

File: packages/uns-py/src/uns_kit/uns_mqtt_proxy.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Any, Dict, Optional

from .client import UnsMqttClient
from .packet import UnsPacket, isoformat
from .proxy import UnsProxy
from .topic_builder import TopicBuilder


class MessageMode(str, Enum):
    RAW = "raw"
    DELTA = "delta"
    BOTH = "both"


@dataclass
class LastValueEntry:
    value: Any
    uom: Optional[str]
    timestamp: datetime


class UnsMqttProxy(UnsProxy):
    def __init__(
        self,
        host: str,
        *,
        process_name: str,
        instance_name: str,
        package_name: str = "uns-kit",
        package_version: str = "0.0.1",
        port: Optional[int] = None,
        username: Optional[str] = None,
        password: Optional[str] = None,
        tls: bool = False,
        client_id: Optional[str] = None,
        keepalive: int = 60,
        clean_session: bool = True,
    ) -> None:
        self.topic_builder = TopicBuilder(package_name, package_version, process_name)
        self.instance_status_topic = self.topic_builder.instance_status_topic(instance_name)
        self.client = UnsMqttClient(
            host,
            port=port,
            username=username,
            password=password,
            tls=tls,
            client_id=client_id,
            keepalive=keepalive,
            clean_session=clean_session,
            topic_builder=self.topic_builder,
            instance_name=instance_name,
            publisher_active=True,
            subscriber_active=True,
        )
        super().__init__(self.client, self.instance_status_topic, instance_name)
        self._last_values: Dict[str, LastValueEntry] = {}
        self._sequence_ids: Dict[str, int] = {}

    async def connect(self) -> None:
        await self.client.connect()
        await self.start()

    async def close(self) -> None:
        await self.stop()
        await self.client.close()

    async def publish_message(self, topic: str, payload: str | bytes) -> None:
        await self.client.publish_raw(topic, payload)

    async def publish_packet(self, topic: str, packet: Dict[str, Any]) -> None:
        await self.client.publish_packet(topic, packet)

    async def publish_mqtt_message(self, mqtt_message: Dict[str, Any], mode: MessageMode = MessageMode.RAW) -> None:
        attrs = mqtt_message.get("attributes")
        if attrs is None:
            raise ValueError("mqtt_message must include attributes")
        if not isinstance(attrs, list):
            attrs = [attrs]

        base_topic = mqtt_message.get("topic", "")
        asset = mqtt_message.get("asset")
        asset_description = mqtt_message.get("assetDescription")
        object_type = mqtt_message.get("objectType")
        object_type_description = mqtt_message.get("objectTypeDescription")
        object_id = mqtt_message.get("objectId")

        for attr in attrs:
            attribute = attr.get("attribute")
            if attribute is None:
                raise ValueError("attribute is required")
            description = attr.get("description")
            tags = attr.get("tags")
            attribute_needs_persistence = attr.get("attributeNeedsPersistence")

            message = attr.get("message")
            if message is None:
                if "data" in attr:
                    message = {
                        "data": attr["data"],
                        **({"createdAt": attr["createdAt"]} if attr.get("createdAt") else {}),
                        **({"expiresAt": attr["expiresAt"]} if attr.get("expiresAt") else {}),
                    }
                elif "table" in attr:
                    message = {
                        "table": attr["table"],
                        **({"createdAt": attr["createdAt"]} if attr.get("createdAt") else {}),
                        **({"expiresAt": attr["expiresAt"]} if attr.get("expiresAt") else {}),
                    }
                else:
                    raise ValueError("Attribute entry must include exactly one of data/table/message")

            packet = UnsPacket.from_message(message)

            msg = {
                "topic": base_topic,
                "asset": asset,
                "assetDescription": asset_description,
                "objectType": object_type,
                "objectTypeDescription": object_type_description,
                "objectId": object_id,
                "attribute": attribute,
                "description": description or attribute,
                "tags": tags,
                "attributeNeedsPersistence": attribute_needs_persistence,
                "packet": packet,
            }

            if mode == MessageMode.RAW:
                await self._process_and_publish(msg, value_is_cumulative=False)
            elif mode == MessageMode.DELTA:
                delta_msg = dict(msg)
                delta_msg["attribute"] = f"{attribute}-delta"
                delta_msg["description"] = f"{msg['description']} (delta)"
                await self._process_and_publish(delta_msg, value_is_cumulative=True)
            elif mode == MessageMode.BOTH:
                await self._process_and_publish(msg, value_is_cumulative=False)
                delta_msg = dict(msg)
                delta_msg["attribute"] = f"{attribute}-delta"
                delta_msg["description"] = f"{msg['description']} (delta)"
                await self._process_and_publish(delta_msg, value_is_cumulative=True)

    def _resolve_object_identity(self, msg: Dict[str, Any]) -> None:
        topic = msg.get("topic", "")
        provided_type = msg.get("objectType")
        provided_id = msg.get("objectId")
        provided_asset = msg.get("asset")

        parts = [p for p in topic.split("/") if p]
        parsed_type = parts[-2] if len(parts) >= 2 else None
        parsed_id = parts[-1] if len(parts) >= 1 else None
        parsed_asset = parts[-3] if len(parts) >= 3 else None

        msg["objectType"] = provided_type or parsed_type
        msg["objectId"] = provided_id or parsed_id or "main"
        msg["asset"] = provided_asset or parsed_asset

    def _normalize_topic(self, topic: str) -> str:
        return topic if topic.endswith("/") else f"{topic}/"

    async def _process_and_publish(self, msg: Dict[str, Any], *, value_is_cumulative: bool) -> None:
        self._resolve_object_identity(msg)
        base_topic = self._normalize_topic(msg.get("topic", ""))

        packet = msg["packet"]
        message = packet.get("message", {})
        data = message.get("data")
        table = message.get("table")

        attribute_type = "Data" if data is not None else "Table" if table is not None else None
        data_group = ""
        if isinstance(data, dict):
            data_group = data.get("dataGroup") or ""
        if isinstance(table, dict):
            data_group = table.get("dataGroup") or ""

        await self.register_unique_topic(
            {
                "timestamp": isoformat(datetime.utcnow()),
                "topic": base_topic,
                "asset": msg.get("asset"),
                "assetDescription": msg.get("assetDescription"),
                "objectType": msg.get("objectType"),
                "objectTypeDescription": msg.get("objectTypeDescription"),
                "objectId": msg.get("objectId"),
                "attribute": msg.get("attribute"),
                "attributeType": attribute_type,
                "description": msg.get("description"),
                "tags": msg.get("tags"),
                "attributeNeedsPersistence": msg.get("attributeNeedsPersistence"),
                "dataGroup": data_group,
            }
        )

        publish_topic = (
            f"{base_topic}"
            f"{msg.get('asset') + '/' if msg.get('asset') else ''}"
            f"{msg.get('objectType') + '/' if msg.get('objectType') else ''}"
            f"{msg.get('objectId') + '/' if msg.get('objectId') else ''}"
            f"{msg.get('attribute')}"
        )

        seq_id = self._sequence_ids.get(base_topic, 0)
        self._sequence_ids[base_topic] = seq_id + 1
        packet["sequenceId"] = seq_id

        if isinstance(data, dict):
            time_value = data.get("time")
            if not time_value:
                time_value = UnsPacket.data(value=0)["message"]["data"]["time"]
                data["time"] = time_value
            current_time = datetime.fromisoformat(time_value.replace("Z", "+00:00"))
            new_value = data.get("value")
            new_uom = data.get("uom")
            last = self._last_values.get(publish_topic)
            if last:
                interval_ms = int((current_time - last.timestamp).total_seconds() * 1000)
                packet["interval"] = interval_ms
                self._last_values[publish_topic] = LastValueEntry(new_value, new_uom, current_time)
                if value_is_cumulative and isinstance(new_value, (int, float)) and isinstance(last.value, (int, float)):
                    if new_value == 0:
                        return
                    delta = new_value - last.value
                    data["value"] = new_value if delta < 0 else delta
                await self.client.publish_raw(publish_topic, UnsPacket.to_json(packet))
            else:
                self._last_values[publish_topic] = LastValueEntry(new_value, new_uom, current_time)
                if not value_is_cumulative:
                    await self.client.publish_raw(publish_topic, UnsPacket.to_json(packet))
        elif isinstance(table, dict):
            await self.client.publish_raw(publish_topic, UnsPacket.to_json(packet))
        else:
            raise ValueError("packet.message must include data or table")
